src/clang_parser.cpp

```diff
@@ -592,8 +592,11 @@ bool ClangParser::parse(ast::Program *program, BPFtrace &bpftrace, std::vector<s
     args.push_back(flag.c_str());
   }
 
-  bool process_btf = program->c_definitions.empty() ||
-                     (bpftrace.force_btf_ && bpftrace.btf_.has_data());
+  // Don't parse BTF for programs containing userspace probes if the user
+  // defined some custom data types, since those may conflict with BTF types.
+  bool process_btf = bpftrace.btf_.has_data() &&
+                     (!program->has_userspace_probes() ||
+                      program->c_definitions.empty());
```

Comment:
I found that this could cause the redefinition errors. For example:

```
% sudo ./src/bpftrace ../tools/tcplife.bt
/lib/modules/5.9.0-rc6+/source/include/net/tcp_states.h:13:2: error: redefinition of enumerator 'TCP_ESTABLISHED'
/bpftrace/include/__btf_generated_header.h:9418:2: note: previous definition is here
/lib/modules/5.9.0-rc6+/source/include/net/tcp_states.h:14:2: error: redefinition of enumerator 'TCP_SYN_SENT'
/bpftrace/include/__btf_generated_header.h:9419:2: note: previous definition is here
[...]
```

We need to think of another condition to solve this.

Reply:
Ouch, not very nice to break a tool we ship. I think we should revert this PR until we can make it work in all cases.

Continuing discussion from https://github.com/iovisor/bpftrace/pull/1506#issuecomment-690995775 , @viktormalik , do you think you could take a look at option 1?

Reply:
Yeah, I'll look into that. It seems to be the cleanest solution and looks like it has use cases, after all.

Reply:
So, after some experiments, I discovered a problem with this approach  - when asking BTF to dump a type definition, it also dumps definitions of all other types needed for successful compilation. If one of those types is overridden by the user, a re-definition error occurs (and it's quite difficult to discover which types were dumped).

This leads me to another thought - do we really want to allow user to do this? I'd say that allowing overriding a part of kernel types while taking the rest from BTF will always eventually blow up somewhere. I'd suggest enabling only the following scenarios:
- there are no user-defined types -> use BTF
- there are user-defined types but they do not override kernel types -> use BTF
- there are user-defined types conflicting with kernel types -> don't use BTF

This should allow user to define his own types that use existing kernel types without any problems, but if he wants to use types with conflicting names (e.g. for userspace tracing), he must define/include all the necessary types manually (or use them from `linux/types.h`, which is included in such case). This approach should be also backwards compatible for scripts that include kernel headers (such as the above `tcplife.bt`) as those must have already specified all the necessary types.

Any thoughs before I start implementing this?

Reply:
I've moved the discussion to #1511